Index: xrdp-0.7.0/common/os_calls.c
===================================================================
--- xrdp-0.7.0.orig/common/os_calls.c	2012-12-19 15:15:14.000000000 +0100
+++ xrdp-0.7.0/common/os_calls.c	2012-12-19 21:05:41.000000000 +0100
@@ -107,21 +107,21 @@
     {
         if (app_name[0] != 0)
         {
-            if (!g_directory_exist("/tmp/.xrdp"))
+            if (!g_directory_exist("/var/run/xrdp"))
             {
-                if (!g_create_dir("/tmp/.xrdp"))
+                if (!g_create_dir("/var/run/xrdp"))
                 {
                     printf("g_mk_temp_dir: g_create_dir failed\n");
                     return 1;
                 }
 
-                g_chmod_hex("/tmp/.xrdp", 0x1777);
+                g_chmod_hex("/var/run/xrdp", 0x1777);
             }
 
             snprintf(g_temp_base, sizeof(g_temp_base),
-                     "/tmp/.xrdp/%s-XXXXXX", app_name);
+                     "/var/run/xrdp/%s-XXXXXX", app_name);
             snprintf(g_temp_base_org, sizeof(g_temp_base_org),
-                     "/tmp/.xrdp/%s-XXXXXX", app_name);
+                     "/var/run/xrdp/%s-XXXXXX", app_name);
 
             if (mkdtemp(g_temp_base) == 0)
             {
Index: xrdp-0.7.0/sesman/chansrv/chansrv.c
===================================================================
--- xrdp-0.7.0.orig/sesman/chansrv/chansrv.c	2012-12-19 11:16:57.000000000 +0100
+++ xrdp-0.7.0/sesman/chansrv/chansrv.c	2012-12-19 21:07:31.000000000 +0100
@@ -814,7 +814,7 @@
     if (g_use_unix_socket)
     {
         g_lis_trans = trans_create(2, 8192, 8192);
-        g_snprintf(port, 255, "/tmp/.xrdp/xrdp_chansrv_socket_%d",
+        g_snprintf(port, 255, "/var/run/xrdp/xrdp_chansrv_socket_%d",
                    7200 + g_display_num);
     }
     else
Index: xrdp-0.7.0/sesman/sessvc/sessvc.c
===================================================================
--- xrdp-0.7.0.orig/sesman/sessvc/sessvc.c	2012-12-19 11:16:57.000000000 +0100
+++ xrdp-0.7.0/sesman/sessvc/sessvc.c	2012-12-19 21:08:22.000000000 +0100
@@ -55,14 +55,14 @@
 {
     char text[256];
 
-    g_snprintf(text, 255, "/tmp/.xrdp/xrdp_chansrv_%8.8x_main_term", pid);
+    g_snprintf(text, 255, "/var/run/xrdp/xrdp_chansrv_%8.8x_main_term", pid);
 
     if (g_file_exist(text))
     {
         g_file_delete(text);
     }
 
-    g_snprintf(text, 255, "/tmp/.xrdp/xrdp_chansrv_%8.8x_thread_done", pid);
+    g_snprintf(text, 255, "/var/run/xrdp/xrdp_chansrv_%8.8x_thread_done", pid);
 
     if (g_file_exist(text))
     {
Index: xrdp-0.7.0/xrdp/xrdp_mm.c
===================================================================
--- xrdp-0.7.0.orig/xrdp/xrdp_mm.c	2012-12-19 11:16:57.000000000 +0100
+++ xrdp-0.7.0/xrdp/xrdp_mm.c	2012-12-19 21:09:21.000000000 +0100
@@ -458,7 +458,7 @@
 
                 if (use_uds)
                 {
-                    g_snprintf(text, 255, "/tmp/.xrdp/xrdp_display_%d", self->display);
+                    g_snprintf(text, 255, "/var/run/xrdp/xrdp_display_%d", self->display);
                 }
                 else
                 {
@@ -873,7 +873,7 @@
                 /* connect channel redir */
                 if ((ip == 0) || (g_strcmp(ip, "127.0.0.1") == 0) || (ip[0] == 0))
                 {
-                    g_snprintf(port, 255, "/tmp/.xrdp/xrdp_chansrv_socket_%d", 7200 + display);
+                    g_snprintf(port, 255, "/var/run/xrdp/xrdp_chansrv_socket_%d", 7200 + display);
                 }
                 else
                 {
Index: xrdp-0.7.0/xrdpapi/xrdpapi.c
===================================================================
--- xrdp-0.7.0.orig/xrdpapi/xrdpapi.c	2012-12-19 11:16:57.000000000 +0100
+++ xrdp-0.7.0/xrdpapi/xrdpapi.c	2012-12-19 21:10:05.000000000 +0100
@@ -135,7 +135,7 @@
     memset(&s, 0, sizeof(struct sockaddr_un));
     s.sun_family = AF_UNIX;
     bytes = sizeof(s.sun_path);
-    snprintf(s.sun_path, bytes - 1, "/tmp/.xrdp/xrdpapi_%d", wts->display_num);
+    snprintf(s.sun_path, bytes - 1, "/var/run/xrdp/xrdpapi_%d", wts->display_num);
     s.sun_path[bytes - 1] = 0;
     bytes = sizeof(struct sockaddr_un);
 
