Index: xrdp-0.7.0/xorg/X11R7.6/rdp/Makefile
===================================================================
--- xrdp-0.7.0.orig/xorg/X11R7.6/rdp/Makefile	2013-03-29 18:18:24.000000000 +0100
+++ xrdp-0.7.0/xorg/X11R7.6/rdp/Makefile	2013-04-01 15:11:28.627658942 +0200
@@ -1,11 +1,13 @@
 
 #X11RDPBASE is an environment variable that needs to be set
 
+XSRCBASE = ../../../../xorg-server-1.12.4
+
+X11RDPBASE = $(XSRCBASE)/build-main
+
 INCBASE = $(X11RDPBASE)/include
 LIBBASE = $(X11RDPBASE)/lib
 
-XSRCBASE = ../build_dir/xorg-server-1.9.3
-
 OBJS = rdpmain.o rdpdraw.o rdpinput.o rdpmisc.o rdpup.o rdprandr.o \
 rdpCopyArea.o rdpPolyFillRect.o rdpPutImage.o rdpPolyRectangle.o \
 rdpPolylines.o rdpPolySegment.o rdpFillSpans.o rdpSetSpans.o \
@@ -21,38 +23,57 @@
 #fbcmap_mi.o
 #fbcmap.o
 
-LIBS = $(XSRCBASE)/dbe/.libs/libdbe.a \
-	$(XSRCBASE)/dix/.libs/libdix.a \
-	$(XSRCBASE)/dix/.libs/libmain.a \
-	$(XSRCBASE)/fb/.libs/libfb.a \
-	$(XSRCBASE)/mi/.libs/libmi.a \
-	$(XSRCBASE)/os/.libs/libos.a \
-	$(XSRCBASE)/randr/.libs/librandr.a \
-	$(XSRCBASE)/record/.libs/librecord.a \
-	$(XSRCBASE)/render/.libs/librender.a \
-	$(XSRCBASE)/xkb/.libs/libxkb.a \
-	$(XSRCBASE)/Xext/.libs/libXext.a \
-	$(XSRCBASE)/Xi/.libs/libXi.a \
-	$(XSRCBASE)/glx/.libs/libglx.a \
-	$(XSRCBASE)/xfixes/.libs/libxfixes.a \
-	$(XSRCBASE)/damageext/.libs/libdamageext.a \
-	$(XSRCBASE)/miext/damage/.libs/libdamage.a \
-	librdp.a
+LIBS = $(X11RDPBASE)/composite/.libs/libcomposite.a \
+       $(X11RDPBASE)/config/.libs/libconfig.a \
+       $(X11RDPBASE)/damageext/.libs/libdamageext.a \
+       $(X11RDPBASE)/dbe/.libs/libdbe.a \
+       $(X11RDPBASE)/dix/.libs/libdix.a \
+       $(X11RDPBASE)/dix/.libs/libmain.a \
+       $(X11RDPBASE)/exa/.libs/libexa.a \
+       $(X11RDPBASE)/fb/.libs/libfb.a \
+       $(X11RDPBASE)/glx/.libs/libglx.a \
+       $(X11RDPBASE)/glx/.libs/libglxdri.a \
+       $(X11RDPBASE)/miext/cw/.libs/libcw.a \
+       $(X11RDPBASE)/miext/damage/.libs/libdamage.a \
+       $(X11RDPBASE)/miext/shadow/.libs/libshadow.a \
+       $(X11RDPBASE)/mi/.libs/libmi.a \
+       $(X11RDPBASE)/os/.libs/libos.a \
+       $(X11RDPBASE)/randr/.libs/librandr.a \
+       $(X11RDPBASE)/record/.libs/librecord.a \
+       $(X11RDPBASE)/render/.libs/librender.a \
+       $(X11RDPBASE)/xkb/.libs/libxkb.a \
+       $(X11RDPBASE)/Xext/.libs/libXextbuiltin.a \
+       $(X11RDPBASE)/Xext/.libs/libXext.a \
+       $(X11RDPBASE)/Xext/.libs/libXextmodule.a \
+       $(X11RDPBASE)/xfixes/.libs/libxfixes.a \
+       $(X11RDPBASE)/Xi/.libs/libXi.a \
+       $(X11RDPBASE)/xkb/.libs/libxkb.a \
+       $(X11RDPBASE)/miext/sync/.libs/libsync.a \
+        librdp.a
 
 LLIBS = -Wl,-rpath=$(LIBBASE) -lfreetype -lz -lm -lXfont -lXau \
-	-lXdmcp -lpixman-1 -lrt -ldl -lcrypto -lGL -lXdamage
+        -lXdmcp -lpixman-1 -lrt -ldl -lcrypto -lGL -lXdamage -lgcrypt
 
 CFLAGS = -O2 -Wall -fno-strength-reduce \
+	-march=native -g \
 	-I../../include \
 	-I../../cfb \
 	-I../../mfb \
 	-I../../mi \
+	-I$(XSRCBASE)/include \
+	-I$(XSRCBASE)/Xext \
+	-I$(XSRCBASE)/mi \
+	-I$(XSRCBASE)/render \
+	-I$(XSRCBASE)/randr \
+	-I$(XSRCBASE)/fb \
+	-I/usr/include/pixman-1 \
 	-I$(INCBASE) \
 	-I$(INCBASE)/X11 \
 	-I$(INCBASE)/X11/fonts \
 	-I$(INCBASE)/X11/extensions \
 	-I$(INCBASE)/xorg \
 	-I$(INCBASE)/pixman-1 \
+	-I../xrdp/common \
 	-I../../os \
 	-I../../render \
 	-I../xfree86/common \
@@ -70,7 +91,8 @@
 	-DBIGREQS -D_POSIX_SOURCE -D_BSD_SOURCE -DSCREENSAVER \
 	-DXCMISC -DXTEST -DXTRAP -DXV -DXSYNC -DMITMISC -DPANORAMIX \
 	-DDPMSExtension -DXvExtension -DXvMCExtension -DXResExtension \
-	-DMITSHM -DPIXPRIV -DNDEBUG -DDDXOSINIT -DXKB -DXINPUT
+	-DMITSHM -DPIXPRIV -DNDEBUG -DDDXOSINIT -DXKB -DXINPUT \
+	-DHAVE_STRNDUP -DSECURITY -DCOMPOSITE
 
 #-DDAMAGE
 
@@ -89,7 +111,7 @@
 # these are not needed I think
 # -Dlinux -D__i386__
 
-LDFLAGS = -L$(LIBBASE) -Wl,-rpath,$(LIBBASE)
+#LDFLAGS = -L$(LIBBASE) -Wl,-rpath,$(LIBBASE)
 
 all: X11rdp
 
@@ -101,11 +123,11 @@
 clean:
 	rm -f $(OBJS) librdp.a
 
-miinitext.o: ../build_dir/xorg-server-1.9.3/mi/miinitext.c Makefile
-	$(CC) $(CFLAGS) -I../build_dir/xorg-server-1.9.3/Xext -c ../build_dir/xorg-server-1.9.3/mi/miinitext.c
+miinitext.o: $(XSRCBASE)/mi/miinitext.c Makefile
+	$(CC) $(CFLAGS) -I$(XSRCBASE)/Xext -c $(XSRCBASE)/mi/miinitext.c
 
-fbcmap.o: ../build_dir/xorg-server-1.9.3/fb/fbcmap.c
-	$(CC) $(CFLAGS) -c ../build_dir/xorg-server-1.9.3/fb/fbcmap.c
+fbcmap.o: $(XSRCBASE)/fb/fbcmap.c
+	$(CC) $(CFLAGS) -c $(XSRCBASE)/fb/fbcmap.c
 
-fbcmap_mi.o: ../build_dir/xorg-server-1.9.3/fb/fbcmap_mi.c
-	$(CC) $(CFLAGS) -c ../build_dir/xorg-server-1.9.3/fb/fbcmap_mi.c
+fbcmap_mi.o: $(XSRCBASE)/fb/fbcmap_mi.c
+	$(CC) $(CFLAGS) -c $(XSRCBASE)/fb/fbcmap_mi.c
Index: xrdp-0.7.0/xorg/X11R7.6/rdp/rdp.h
===================================================================
--- xrdp-0.7.0.orig/xorg/X11R7.6/rdp/rdp.h	2013-04-01 14:51:11.000000000 +0200
+++ xrdp-0.7.0/xorg/X11R7.6/rdp/rdp.h	2013-04-01 15:11:28.627658942 +0200
@@ -23,6 +23,7 @@
 #define __arm32__
 #endif
 
+#include <X11/Xfuncproto.h>
 #include "xorg-server.h"
 #include <stdio.h>
 #include <stdarg.h>
@@ -36,10 +37,10 @@
 #include <netinet/tcp.h>
 #include <netinet/in.h>
 #include <netdb.h>
-#include "X.h"
+#include <X11/X.h>
 #define NEED_EVENTS
-#include "Xproto.h"
-#include "Xos.h"
+#include <X11/Xproto.h>
+#include <X11/Xos.h>
 #include "scrnintstr.h"
 #include "servermd.h"
 #define PSZ 8
@@ -53,20 +54,21 @@
 #include "mipointer.h"
 #include "dixstruct.h"
 #include "propertyst.h"
-#include "Xatom.h"
+#include <X11/Xatom.h>
 #include "dix.h"
 #include "X11/keysym.h"
 #include "dixfontstr.h"
-#include "fontstruct.h"
+#include <X11/fonts/fontstruct.h>
 #include "cursorstr.h"
 #include "picturestr.h"
-#include "XKBstr.h"
+#include <X11/extensions/XKBstr.h>
 #include "inputstr.h"
 #include "randrstr.h"
 #include "mi.h"
 #include "fb.h"
 #include "micmap.h"
 #include "events.h"
+#include "eventstr.h"
 #include "exevents.h"
 #include "xserver-properties.h"
 #include "xkbsrv.h"
@@ -145,7 +147,7 @@
   CompositeProcPtr Composite;
   GlyphsProcPtr Glyphs;
   /* Backing store procedures */
-  RestoreAreasProcPtr RestoreAreas;
+  //RestoreAreasProcPtr RestoreAreas;
 
   int rdp_width;
   int rdp_height;
@@ -450,7 +454,7 @@
 void
 rdpSpriteDeviceCursorCleanup(DeviceIntPtr pDev, ScreenPtr pScr);
 void
-PtrAddEvent(int buttonMask, int x, int y);
+PtrAddEvent(int buttonMask, int mode, double x, double y);
 void
 KbdAddEvent(int down, int param1, int param2, int param3, int param4);
 void
Index: xrdp-0.7.0/xorg/X11R7.6/rdp/rdpdraw.c
===================================================================
--- xrdp-0.7.0.orig/xorg/X11R7.6/rdp/rdpdraw.c	2013-03-25 19:58:04.000000000 +0100
+++ xrdp-0.7.0/xorg/X11R7.6/rdp/rdpdraw.c	2013-04-01 15:11:28.627658942 +0200
@@ -399,7 +399,7 @@
     //pScreen->PaintWindowBorder = g_rdpScreen.PaintWindowBorder;
     pScreen->CopyWindow = g_rdpScreen.CopyWindow;
     pScreen->ClearToBackground = g_rdpScreen.ClearToBackground;
-    pScreen->RestoreAreas = g_rdpScreen.RestoreAreas;
+    //pScreen->RestoreAreas = g_rdpScreen.RestoreAreas;
     return 1;
 }
 
@@ -1135,8 +1135,8 @@
     LLOGLN(0, ("in rdpRestoreAreas"));
     RegionInit(&reg, NullBox, 0);
     RegionCopy(&reg, prgnExposed);
-    g_pScreen->RestoreAreas = g_rdpScreen.RestoreAreas;
-    rv = g_pScreen->RestoreAreas(pWin, prgnExposed);
+    //g_pScreen->RestoreAreas = g_rdpScreen.RestoreAreas;
+    //rv = g_pScreen->RestoreAreas(pWin, prgnExposed);
 
     if (g_do_dirty_ons)
     {
@@ -1156,8 +1156,9 @@
     }
 
     RegionUninit(&reg);
-    g_pScreen->RestoreAreas = rdpRestoreAreas;
-    return rv;
+    //g_pScreen->RestoreAreas = rdpRestoreAreas;
+    //return rv;
+    return prgnExposed;
 }
 
 /******************************************************************************/
Index: xrdp-0.7.0/xorg/X11R7.6/rdp/rdpinput.c
===================================================================
--- xrdp-0.7.0.orig/xorg/X11R7.6/rdp/rdpinput.c	2013-04-01 14:51:11.000000000 +0200
+++ xrdp-0.7.0/xorg/X11R7.6/rdp/rdpinput.c	2013-04-01 15:11:28.631658943 +0200
@@ -736,13 +736,12 @@
 
 /******************************************************************************/
 static void
-rdpEnqueueMotion(int x, int y)
+rdpEnqueueMotion(int mode, double x, double y)
 {
     int i;
     int n;
-    int valuators[2];
-    EventListPtr rdp_events;
-    xEvent *pev;
+    InternalEvent* rdp_events;
+    ValuatorMask *mask = valuator_mask_new(2);
 
 # if 0
 
@@ -754,20 +753,21 @@
     }
 
 #endif
-    miPointerSetPosition(g_pointer, &x, &y);
-    valuators[0] = x;
-    valuators[1] = y;
+    miPointerSetPosition(g_pointer, mode, &x, &y);
+    valuator_mask_set_double(mask, 0, x);
+    valuator_mask_set_double(mask, 1, y);
 
-    GetEventList(&rdp_events);
+    rdp_events = InitEventList(GetMaximumEventsNum() + 1);
     n = GetPointerEvents(rdp_events, g_pointer, MotionNotify, 0,
-                         POINTER_ABSOLUTE | POINTER_SCREEN,
-                         0, 2, valuators);
+                         POINTER_ABSOLUTE | POINTER_SCREEN, mask);
 
     for (i = 0; i < n; i++)
     {
-        pev = (rdp_events + i)->event;
-        mieqEnqueue(g_pointer, (InternalEvent *)pev);
+        mieqEnqueue(g_pointer, &rdp_events[i]);
     }
+
+    valuator_mask_free(&mask);
+    FreeEventList(rdp_events, GetMaximumEventsNum());
 }
 
 /******************************************************************************/
@@ -776,17 +776,17 @@
 {
     int i;
     int n;
-    EventListPtr rdp_events;
-    xEvent *pev;
+    InternalEvent* rdp_events;
 
-    i = GetEventList(&rdp_events);
-    n = GetPointerEvents(rdp_events, g_pointer, type, buttons, 0, 0, 0, 0);
+    rdp_events = InitEventList(GetMaximumEventsNum() + 1);
+    n = GetPointerEvents(rdp_events, g_pointer, type, buttons, 0, NULL);
 
     for (i = 0; i < n; i++)
     {
-        pev = (rdp_events + i)->event;
-        mieqEnqueue(g_pointer, (InternalEvent *)pev);
+        mieqEnqueue(g_pointer, &rdp_events[i]);
     }
+
+    FreeEventList(rdp_events, GetMaximumEventsNum());
 }
 
 /******************************************************************************/
@@ -795,28 +795,28 @@
 {
     int i;
     int n;
-    EventListPtr rdp_events;
-    xEvent *pev;
+    InternalEvent* rdp_events;
 
-    i = GetEventList(&rdp_events);
-    n = GetKeyboardEvents(rdp_events, g_keyboard, type, scancode);
+    rdp_events = InitEventList(GetMaximumEventsNum() + 1);
+    n = GetKeyboardEvents(rdp_events, g_keyboard, type, scancode, NULL);
 
     for (i = 0; i < n; i++)
     {
-        pev = (rdp_events + i)->event;
-        mieqEnqueue(g_keyboard, (InternalEvent *)pev);
+        mieqEnqueue(g_keyboard, &rdp_events[i]);
     }
+
+    FreeEventList(rdp_events, GetMaximumEventsNum());
 }
 
 /******************************************************************************/
 void
-PtrAddEvent(int buttonMask, int x, int y)
+PtrAddEvent(int buttonMask, int mode, double x, double y)
 {
     int i;
     int type;
     int buttons;
 
-    rdpEnqueueMotion(x, y);
+    rdpEnqueueMotion(mode, x, y);
 
     for (i = 0; i < 5; i++)
     {
Index: xrdp-0.7.0/xorg/X11R7.6/rdp/rdpmain.c
===================================================================
--- xrdp-0.7.0.orig/xorg/X11R7.6/rdp/rdpmain.c	2013-03-25 19:58:04.000000000 +0100
+++ xrdp-0.7.0/xorg/X11R7.6/rdp/rdpmain.c	2013-04-01 15:11:28.631658943 +0200
@@ -385,7 +385,7 @@
     g_rdpScreen.ClearToBackground = pScreen->ClearToBackground;
 
     /* Backing store procedures */
-    g_rdpScreen.RestoreAreas = pScreen->RestoreAreas;
+    //g_rdpScreen.RestoreAreas = pScreen->RestoreAreas;
     g_rdpScreen.WakeupHandler = pScreen->WakeupHandler;
 
     g_rdpScreen.CreateColormap = pScreen->CreateColormap;
@@ -439,7 +439,7 @@
     pScreen->ClearToBackground = rdpClearToBackground;
 
     /* Backing store procedures */
-    pScreen->RestoreAreas = rdpRestoreAreas;
+    //pScreen->RestoreAreas = rdpRestoreAreas;
 
 #if 0
     pScreen->CreateColormap = rdpCreateColormap;
@@ -691,7 +691,7 @@
 
 /******************************************************************************/
 void
-ddxGiveUp(void)
+ddxGiveUp(enum ExitCode error)
 {
     char unixSocketName[128];
 
@@ -736,9 +736,9 @@
 
 /******************************************************************************/
 void
-AbortDDX(void)
+AbortDDX(enum ExitCode error)
 {
-    ddxGiveUp();
+    ddxGiveUp(error);
 }
 
 /******************************************************************************/
Index: xrdp-0.7.0/xorg/X11R7.6/rdp/rdpup.c
===================================================================
--- xrdp-0.7.0.orig/xorg/X11R7.6/rdp/rdpup.c	2013-04-01 14:51:11.000000000 +0200
+++ xrdp-0.7.0/xorg/X11R7.6/rdp/rdpup.c	2013-04-01 15:11:28.631658943 +0200
@@ -670,47 +670,47 @@
                    past the width or height */
                 g_cursor_x = l_bound_by(param1, 0, g_rdpScreen.width - 2);
                 g_cursor_y = l_bound_by(param2, 0, g_rdpScreen.height - 2);
-                PtrAddEvent(g_button_mask, g_cursor_x, g_cursor_y);
+                PtrAddEvent(g_button_mask, Absolute, g_cursor_x, g_cursor_y);
                 break;
             case 101:
                 g_button_mask = g_button_mask & (~1);
-                PtrAddEvent(g_button_mask, g_cursor_x, g_cursor_y);
+                PtrAddEvent(g_button_mask, Absolute, g_cursor_x, g_cursor_y);
                 break;
             case 102:
                 g_button_mask = g_button_mask | 1;
-                PtrAddEvent(g_button_mask, g_cursor_x, g_cursor_y);
+                PtrAddEvent(g_button_mask, Absolute, g_cursor_x, g_cursor_y);
                 break;
             case 103:
                 g_button_mask = g_button_mask & (~4);
-                PtrAddEvent(g_button_mask, g_cursor_x, g_cursor_y);
+                PtrAddEvent(g_button_mask, Absolute, g_cursor_x, g_cursor_y);
                 break;
             case 104:
                 g_button_mask = g_button_mask | 4;
-                PtrAddEvent(g_button_mask, g_cursor_x, g_cursor_y);
+                PtrAddEvent(g_button_mask, Absolute, g_cursor_x, g_cursor_y);
                 break;
             case 105:
                 g_button_mask = g_button_mask & (~2);
-                PtrAddEvent(g_button_mask, g_cursor_x, g_cursor_y);
+                PtrAddEvent(g_button_mask, Absolute, g_cursor_x, g_cursor_y);
                 break;
             case 106:
                 g_button_mask = g_button_mask | 2;
-                PtrAddEvent(g_button_mask, g_cursor_x, g_cursor_y);
+                PtrAddEvent(g_button_mask, Absolute, g_cursor_x, g_cursor_y);
                 break;
             case 107:
                 g_button_mask = g_button_mask & (~8);
-                PtrAddEvent(g_button_mask, g_cursor_x, g_cursor_y);
+                PtrAddEvent(g_button_mask, Absolute, g_cursor_x, g_cursor_y);
                 break;
             case 108:
                 g_button_mask = g_button_mask | 8;
-                PtrAddEvent(g_button_mask, g_cursor_x, g_cursor_y);
+                PtrAddEvent(g_button_mask, Absolute, g_cursor_x, g_cursor_y);
                 break;
             case 109:
                 g_button_mask = g_button_mask & (~16);
-                PtrAddEvent(g_button_mask, g_cursor_x, g_cursor_y);
+                PtrAddEvent(g_button_mask, Absolute, g_cursor_x, g_cursor_y);
                 break;
             case 110:
                 g_button_mask = g_button_mask | 16;
-                PtrAddEvent(g_button_mask, g_cursor_x, g_cursor_y);
+                PtrAddEvent(g_button_mask, Absolute, g_cursor_x, g_cursor_y);
                 break;
             case 200:
                 rdpup_begin_update();
Index: xrdp-0.7.0/xup/xup.c
===================================================================
--- xrdp-0.7.0.orig/xup/xup.c	2013-04-01 14:51:11.000000000 +0200
+++ xrdp-0.7.0/xup/xup.c	2013-04-01 15:11:28.631658943 +0200
@@ -19,6 +19,11 @@
  */
 
 #include "xup.h"
+#include "log.h"
+#include <stdlib.h>
+#include <errno.h>
+#include <unistd.h>
+#include <sys/types.h>
 
 /******************************************************************************/
 /* returns error */
@@ -180,10 +185,12 @@
     {
         if (use_uds)
         {
+            mod->server_msg(mod, "Use local sockets", 0);
             mod->sck = g_tcp_local_socket();
         }
         else
         {
+            mod->server_msg(mod, "Use tcp sockets", 0);
             mod->sck = g_tcp_socket();
             g_tcp_set_non_blocking(mod->sck);
             g_tcp_set_no_delay(mod->sck);
@@ -202,6 +209,7 @@
 
         if (error == -1)
         {
+            log_message(LOG_LEVEL_INFO, "Error during connect err=%d on sck=%d and port=%s for uid=%d", errno,  mod->sck, con_port, getuid());
             if (g_tcp_last_error_would_block(mod->sck))
             {
                 error = 0;
