Index: xrdp-0.7.0/libxrdp/xrdp_orders.c
===================================================================
--- xrdp-0.7.0.orig/libxrdp/xrdp_orders.c	2013-02-07 22:22:33.903178564 +0100
+++ xrdp-0.7.0/libxrdp/xrdp_orders.c	2013-04-01 15:56:27.112871466 +0200
@@ -1872,9 +1872,9 @@
             if (Bpp == 3)
             {
                 pixel = GETPIXEL32(data, j, i, width);
-                out_uint8(self->out_s, pixel >> 16);
-                out_uint8(self->out_s, pixel >> 8);
                 out_uint8(self->out_s, pixel);
+                out_uint8(self->out_s, pixel >> 8);
+                out_uint8(self->out_s, pixel >> 16);
             }
             else if (Bpp == 2)
             {
@@ -2093,9 +2093,9 @@
             if (Bpp == 3)
             {
                 pixel = GETPIXEL32(data, j, i, width);
-                out_uint8(self->out_s, pixel >> 16);
-                out_uint8(self->out_s, pixel >> 8);
                 out_uint8(self->out_s, pixel);
+                out_uint8(self->out_s, pixel >> 8);
+                out_uint8(self->out_s, pixel >> 16);
             }
             else if (Bpp == 2)
             {
