--- a/libxrdp/xrdp_rdp.c
+++ b/libxrdp/xrdp_rdp.c
@@ -1242,7 +1242,7 @@
                 break;
             case 17: /* 17 */
                 DEBUG(("CAPSET_TYPE_OFFSCREEN_CACHE"));
-                xrdp_process_offscreen_bmpcache(self, s, len);
+                //xrdp_process_offscreen_bmpcache(self, s, len);
                 break;
             case RDP_CAPSET_BMPCACHE2: /* 19 */
                 DEBUG(("RDP_CAPSET_BMPCACHE2"));
