Index: xrdp-bartw72/instfiles/Makefile.am
===================================================================
--- xrdp-bartw72.orig/instfiles/Makefile.am	2013-11-23 14:13:04.000000000 +0100
+++ xrdp-bartw72/instfiles/Makefile.am	2013-11-23 14:26:57.000000000 +0100
@@ -29,4 +29,4 @@
 install-data-hook:
 	chmod 755 $(DESTDIR)$(sysconfdir)/xrdp/xrdp.sh
 	chmod 755 $(DESTDIR)$(sysconfdir)/init.d/xrdp
-	sed -i 's|__BASE__|$(prefix)|' $(sysconfdir)/init.d/xrdp;
+	sed -i 's|__BASE__|$(prefix)|' $(DESTDIR)$(sysconfdir)/init.d/xrdp;
