--- a/instfiles/init.d/xrdp
+++ b/instfiles/init.d/xrdp
@@ -135,6 +135,18 @@
         else
             log_progress_msg "apparently not running"
         fi
+        if pidofproc /usr/bin/X11rdp > /dev/null; then
+            sh /usr/sbin/xrdp-killall >/dev/null
+        elif pidofproc /usr/bin/Xvnc > /dev/null; then
+            sh /usr/sbin/xrdp-killall >/dev/null
+        fi
+        if [ -d /tmp/.xrdp ] ; then
+                sleep 3
+                rm -rf /tmp/.xrdp
+        fi
+	if [ -d /tmp/.X11-unix ] ; then
+		rm -f /tmp/.X11-unix/X??
+	fi
         log_end_msg $exitval
 	;;
   force-stop)
--- a/xrdp/xrdp.ini
+++ b/xrdp/xrdp.ini
@@ -3,7 +3,7 @@
 bitmap_cache=yes
 bitmap_compression=yes
 port=3389
-crypt_level=high
+crypt_level=low
 allow_channels=true
 max_bpp=24
 fork=yes
