Index: xrdp-bartw72/sesman/session.c
===================================================================
--- xrdp-bartw72.orig/sesman/session.c	2013-12-06 09:20:34.837415208 +0100
+++ xrdp-bartw72/sesman/session.c	2013-12-06 09:22:50.045294268 +0100
@@ -641,43 +641,50 @@
             }
             else /* parent (child sesman)*/
             {
-                wait_for_xserver(display);
-                g_snprintf(text, 255, "%d", display);
-                g_setenv("XRDP_SESSVC_DISPLAY", text, 1);
-                g_snprintf(text, 255, ":%d.0", display);
-                g_setenv("DISPLAY", text, 1);
-                /* new style waiting for clients */
-                session_start_sessvc(xpid, wmpid, data, username, display);
+				wait_for_xserver(display);
+                if (x_server_running(display))
+                {
+					g_snprintf(text, 255, "%d", display);
+					g_setenv("XRDP_SESSVC_DISPLAY", text, 1);
+					g_snprintf(text, 255, ":%d.0", display);
+					g_setenv("DISPLAY", text, 1);
+					/* new style waiting for clients */
+					session_start_sessvc(xpid, wmpid, data, username, display);
+                }
             }
         }
     }
     else /* parent sesman process */
     {
-        temp->item->pid = pid;
-        temp->item->display = display;
-        temp->item->width = width;
-        temp->item->height = height;
-        temp->item->bpp = bpp;
-        temp->item->data = data;
-        g_strncpy(temp->item->client_ip, client_ip, 255);   /* store client ip data */
-        g_strncpy(temp->item->name, username, 255);
+		wait_for_xserver(display);
+        if (x_server_running(display))
+        {
+			temp->item->pid = pid;
+			temp->item->display = display;
+			temp->item->width = width;
+			temp->item->height = height;
+			temp->item->bpp = bpp;
+			temp->item->data = data;
+			g_strncpy(temp->item->client_ip, client_ip, 255);   /* store client ip data */
+			g_strncpy(temp->item->name, username, 255);
 
-        ltime = g_time1();
-        localtime_r(&ltime, &stime);
-        temp->item->connect_time.year = (tui16)(stime.tm_year + 1900);
-        temp->item->connect_time.month = (tui8)stime.tm_mon;
-        temp->item->connect_time.day = (tui8)stime.tm_mday;
-        temp->item->connect_time.hour = (tui8)stime.tm_hour;
-        temp->item->connect_time.minute = (tui8)stime.tm_min;
-        zero_time(&(temp->item->disconnect_time));
-        zero_time(&(temp->item->idle_time));
+			ltime = g_time1();
+			localtime_r(&ltime, &stime);
+			temp->item->connect_time.year = (tui16)(stime.tm_year + 1900);
+			temp->item->connect_time.month = (tui8)stime.tm_mon;
+			temp->item->connect_time.day = (tui8)stime.tm_mday;
+			temp->item->connect_time.hour = (tui8)stime.tm_hour;
+			temp->item->connect_time.minute = (tui8)stime.tm_min;
+			zero_time(&(temp->item->disconnect_time));
+			zero_time(&(temp->item->idle_time));
 
-        temp->item->type = type;
-        temp->item->status = SESMAN_SESSION_STATUS_ACTIVE;
+			temp->item->type = type;
+			temp->item->status = SESMAN_SESSION_STATUS_ACTIVE;
 
-        temp->next = g_sessions;
-        g_sessions = temp;
-        g_session_count++;
+			temp->next = g_sessions;
+			g_sessions = temp;
+			g_session_count++;
+        }
     }
 
     return display;
