--- debian/rules.orig	2013-09-21 22:22:18.068755250 +0200
+++ debian/rules	2013-09-21 22:15:05.909128970 +0200
@@ -70,7 +70,7 @@
 	config_backend += --disable-config-udev --disable-config-hal
 endif
 
-vars = $(shell DEB_BUILD_MAINT_OPTIONS="hardening=+pie" DEB_LDFLAGS_MAINT_APPEND="-Wl,-Bsymbolic" DEB_CPPFLAGS_MAINT_APPEND="-DPRE_RELEASE=0 $(cppflags)" dpkg-buildflags --export=configure) LIBS="$(libs)"
+vars = $(shell DEB_BUILD_MAINT_OPTIONS="hardening=+pie" DEB_CFLAGS_MAINT_APPEND="-march=native" DEB_LDFLAGS_MAINT_APPEND="-Wl,-Bsymbolic" DEB_CPPFLAGS_MAINT_APPEND="-DPRE_RELEASE=0" dpkg-buildflags --export=configure) LIBS="$(libs)"
 
 ifneq (,$(filter nocheck, $(DEB_BUILD_OPTIONS)))
 	config_tests = --disable-unit-tests
@@ -190,7 +190,7 @@
 	dh_testdir
 	mkdir -p build-$*
 	cd build-$* && \
-	../configure \
+	CC=gcc-4.6 ../configure \
 	  --prefix=/usr \
 	  --mandir=\$${prefix}/share/man \
 	  --infodir=\$${prefix}/share/info \
